proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+- *Union interfaces*: A few interfaces are known by the language and used in its implementation of unions. `IUnion` is used to recognize union types and support union matching. An optional interface, `IUnion<TUnion>`, adds broader support for creating union values from potential case types, and is implemented by the union types arising from union declarations.
+
+## Motivation
+[motivation]: #motivation
+
+Unions are a long-requested C# feature, which allows expressing values from a closed set of types in a way that pattern matching can trust to be exhaustive.
+
+The separation between union *types* and union *declarations* allows C# to have a succinct union declaration syntax with opinionated semantics, while also allowing existing types or types with other implementation choices to opt into union behaviors.
+
+The proposed unions in C# are unions of *types* and not "discriminated" or "tagged". "Discriminated unions" can be expressed in terms of "type unions" by using fresh type declarations as case types.
+
+## Detailed design
+[design]: #detailed-design
+
+### Union interfaces
+
+The following interfaces are recognized by the language in its implementation of union features.
+
+#### Union access interface
+
+The `IUnion` interface marks a type as a union type at compile time and provides a way to access union contents at runtime.
+
+```csharp
+public interface IUnion
+{
+    // The value of the union or null
+    object? Value { get; }
+}
+```
+As we'll see later:
+* The `IUnion` interface is all that is necessary for a type to be considered a union by the language.  
+* The `Value` property of the interface is targeted by the compiler to implement pattern matching.
+* Union declarations are generated by the compiler to implement this interface.
+
+Example use:
+
+```csharp
+if (value is IUnion union && union.Value is null) { ... }
+```
+
+#### Union creation interface
+
+The `IUnion<TUnion>` interface provides a means to create union instances at runtime.
+
+```csharp
+public interface IUnion<TUnion> : IUnion where TUnion : IUnion<TUnion>
+{
+    // Creates a union from a value
+    static abstract bool TryCreate(object? value, [NotNullWhen(true)] out TUnion union);
+}
+```
+
+* Union declarations are generated by the compiler to implement this interface.
+
+Example use:
+
+```csharp
+TUnion ReadUnion<TUnion>() where TUnion : IUnion<TUnion>
+{
+    object val = ReadValue();
+
+    // wrap the value in the union
+    if (TUnion.TryCreate(val, out var union)) 
+        return union;
+
+    throw ...;
+}
+```
+
+### Union types
+
+Any class or struct type that implements the `IUnion` interface is considered a *union type*. For each public constructor with exactly one parameter, the type of that parameter is considered a *case type* of the union type.
```

Comment:
`public` or `accessible` constructor?   I'm fine with the former, just making sure it is intentional.